lib/upnp/igd_service.go

```diff
@@ -49,12 +50,86 @@ type IGDService struct {
 	ServiceID string
 	URL       string
 	URN       string
-	LocalIP   net.IP
+	LocalIPv4 net.IP
+	Interface *net.Interface
+	PinholeID uint16
+}
+
+// TryAddPinhole adds an IPv6 pinhole in accordance to http://upnp.org/specs/gw/UPnP-gw-WANIPv6FirewallControl-v1-Service.pdf
+// This is attempted for each IPv6 on the interface.
+func (s *IGDService) TryAddPinhole(ctx context.Context, protocol nat.Protocol, port int, description string, duration time.Duration) (int, error) {
+	result := 0
+	var returnErr error = nil
+	if s.Interface == nil {
+		return 0, errors.New("no interface")
+	}
+
+	addrs, err := s.Interface.Addrs()
+	if err != nil {
+		return 0, err
+	}
+
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			l.Infoln("Couldn't parse address ", addr, err)
+			continue
+		}
+
+		if ip.To4() == nil && ip.IsGlobalUnicast() {
```

Comment:
We should 
```suggestion
		if ip.To4() != nil || ip.isPrivate() || !ip.IsGlobalUnicast() {
			continue;
		}
```

Otherwise we would request mappings for addresses in the ULA range.

Reply:
You're right, `IsGlobalUnicast` is a bit of a misnomer. I've changed it now